### ISSUE TITLE
[prefer-string-starts-ends-with] False positives when specific length is required

### ISSUE THREAD

**Issue by pgsandstrom** [typescript-eslint/typescript-eslint#896]
**Repro**

```TS
const test = (s1: string, s2: string) => s1.substring(2) === s2
```

This method is suggested to be rewritten with `.endsWith()`. However, that is not feasible. This function requires `s1` to be exactly 2 characters longer than `s2`. That cannot be replicated with `.endsWith()`.

**Expected Result**

There should be no error reported on the above function.

**Actual Result**

An error is reported.

**Additional Info**

I think one could say that the rule fails whenever there is a length requirement hidden in the code. For example, it makes the same error for this function, which requires `s2` to be exactly two characters long to ever return true. Simply replacing it with `endsWith()` would remove all length requirements.

```TS
const anotherTest = (s1: string, s2: string) => s1.substring(s1.length - 2) === s2
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0` |
| `@typescript-eslint/parser`        | `2.0.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.2.1` |
| `node`                             | `12.0.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
This is by design for the rule.
I think if you're intentionally not using `endsWith`, you should disable the lint rule with a comment explaining why you are using the implicit length check.

Personally, if I was reading the rule and saw `s1.substring(2) === s2`, my first thought wouldn't be "s2 needs to be exactly two characters longer than s1", it would instead be "s1 has a 2 character long prefix we don't care about".
From that IMO you should do an explicit length check with `endsWith`, so that your intention is clear.

**Comment by pgsandstrom** [comment]
While the function could be written in many (perhaps better) ways, I don't think my way of writing it is unreasonable or very uncommon, which is what should matter for the purpose of linting.

I think it's problematic that the rule makes a suggestion that would change the behaviour of the code. In some cases, a subtle bug will be introduced into the code. Wouldn't be reasonable that rules that are part of the recommended set is more conservative about things like this?

Of course all of this is subjective, I just wanted to give my two cents.

**Comment by pgsandstrom** [comment]
Fixed in #1920.
